src/Hadolint/Rules.hs

```diff
@@ -1164,3 +1165,19 @@ noIllegalInstructionInOnbuild = instructionRule code severity message check
     check (OnBuild (From _)) = False
     check (OnBuild (Maintainer _)) = False
     check _ = True
+
+
+noSelfreferencingEnv :: Rule
+noSelfreferencingEnv = instructionRule code severity message check
+  where
+    code = "DL3044"
+    severity = ErrorC
+    message = "Do not refer to an environment variable within the same `ENV` statement where it is defined."
+    check (Env pairs) = findReferences pairs
+    check _ = True
+
+    findReferences :: Pairs -> Bool
+    findReferences prs = null [ var | var <- map fst prs,
+                                      var `isSubstringOfAny` map snd prs ]
+    isSubstringOfAny :: Text.Text -> [Text.Text] -> Bool
+    isSubstringOfAny t l = not $ null [ v | v <- l, t `Text.isInfixOf` v]
```

Comment:
Would this trigger on partial matches? For example, I define `FOO` and then do `BAR = ${FOOBAR}`. Would this case trigger the rule incorrectly?

Reply:
Thats a good point. I was thinking about partial matches of the kind `BAR=/somepath/${FOO}`, but forgot your case. Anyways, I fixed it and added tests to cover this case as well.